dev/RadioButtons/RadioButtons.cpp

```diff
@@ -7,136 +7,634 @@
+                            auto const repeaterItemCount = [repeater]()
+                            {
+                                if (auto const itemsSourceView = repeater.ItemsSourceView())
+                                {
+                                    return itemsSourceView.Count();
+                                }
+                                return 0;
+                            }();
+                            // If focus is coming from outside the repeater, put focus on the selected item.
+                            if (oldFocusedRepeaterIndex < 0)
+                            {
+                                if (auto const selectedItem = repeater.TryGetElement(selectedIndex.GetAt(0)))
+                                {
+                                    args.TrySetNewFocusedElement(selectedItem);
+                                }
+                            }
+                            // If focus is coming from the first element to the last element (or last to first), via keyboard we know (suspect)
+                            // that RadioButton has interfered and we want to correct. 
+                            else if (oldFocusedRepeaterIndex == 0 ||
+                                     oldFocusedRepeaterIndex == repeaterItemCount - 1)
+                            {
+                                if (auto const newFocusedElement = args.NewFocusedElement())
+                                {
+                                    if (auto const newFocusedElementAsUIElement = newFocusedElement.try_as<winrt::UIElement>())
+                                    {
+                                        auto const newFocusedRepeaterIndex = repeater.GetElementIndex(newFocusedElementAsUIElement);
+                                        if ((oldFocusedRepeaterIndex == 0 && newFocusedRepeaterIndex == repeaterItemCount - 1) ||
+                                            (oldFocusedRepeaterIndex == repeaterItemCount - 1 && newFocusedRepeaterIndex == 0))
+                                        {
+                                            args.TrySetNewFocusedElement(oldFocusedElement);
+                                        }
+                                    }
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+        }
     }
+}
 
-    UpdateItemsSource();
+void RadioButtons::OnRepeaterLoaded(const winrt::IInspectable&, const winrt::RoutedEventArgs&)
+{
+    if (auto const repeater = m_repeater.get())
+    {
+        if (m_testHooksEnabled)
+        {
+            AttachToLayoutChanged();
+        }
+
+        UpdateSelectedIndex();
+        UpdateSelectedItem();
+        UpdateMaximumColumns();
+        OnRepeaterCollectionChanged(nullptr, nullptr);
+    }
+}
+
+void RadioButtons::KeyDownHandler(const winrt::IInspectable&, const winrt::KeyRoutedEventArgs& args)
+{
```

Comment:
I think you can use this to know the state of the control key instead of figuring it out using KeyDown/KeyUp.
 var isCtrlPressed = Window.Current.CoreWindow.GetAsyncKeyState(VirtualKey.Control).HasFlag(Windows.UI.Core.CoreVirtualKeyStates.Down);
 
@Austin-Lamb - is it ok to call this CoreWindow API to know if the control key is pressed ?

Reply:
While this does work, I'm worried that it is not islands safe.

---
In reply to: [347502892](https://github.com/microsoft/microsoft-ui-xaml/pull/1623#discussion_r347502892) [](ancestors = 347502892)

Reply:
Jevan says it is so sounds good!

---
In reply to: [348665272](https://github.com/microsoft/microsoft-ui-xaml/pull/1623#discussion_r348665272) [](ancestors = 348665272,347502892)